File: frontend/src/lib/api/threatIntel.ts
import { apiGet, apiPatch, apiPost } from './client';
import type { Page } from './chatbot';

export type IOCType = 'ip' | 'domain' | 'url' | 'sha256';
export type IOCSeverity = 'low' | 'medium' | 'high' | 'critical';
export type IOCConfidence = 'low' | 'medium' | 'high';

export interface RiskPoint {
  time: string;
  score: number;
}

export interface ThreatIOC {
  id: string;
  type: IOCType;
  value: string;
  severity: IOCSeverity;
  confidence: IOCConfidence;
  description: string;
  source: string;
  first_seen: string;
  last_seen: string;
  watchlist: boolean;
  tags: string[];
  mitre_techniques: string[];
  risk_timeline: RiskPoint[];
  created_at: string;
  updated_at: string;
}

export interface ThreatIOCCreateInput {
  type: IOCType;
  value: string;
  severity: IOCSeverity;
  confidence: IOCConfidence;
  description?: string;
  source: string;
  first_seen: string;
  last_seen: string;
  watchlist?: boolean;
  tags?: string[];
  mitre_techniques?: string[];
  risk_timeline?: RiskPoint[];
}

export interface ThreatIOCFilters {
  page?: number;
  pageSize?: number;
  search?: string;
  type?: IOCType;
  severity?: IOCSeverity;
  watchlist?: boolean;
}

export interface ThreatIntelSummary {
  total: number;
  critical: number;
  watchlist: number;
  recent_48h: number;
  items: ThreatIOC[];
}

export function listThreatIOCs(filters: ThreatIOCFilters = {}): Promise<Page<ThreatIOC>> {
  const params = new URLSearchParams();
  params.set('page', String(filters.page ?? 1));
  params.set('page_size', String(filters.pageSize ?? 20));
  if (filters.search) params.set('search', filters.search);
  if (filters.type) params.set('type', filters.type);
  if (filters.severity) params.set('severity', filters.severity);
  if (typeof filters.watchlist === 'boolean') params.set('watchlist', String(filters.watchlist));
  return apiGet<Page<ThreatIOC>>(`/api/threat-intel/iocs?${params.toString()}`);
}

export function getThreatIOC(id: string): Promise<ThreatIOC> {
  return apiGet<ThreatIOC>(`/api/threat-intel/iocs/${id}`);
}

export function createThreatIOC(input: ThreatIOCCreateInput): Promise<ThreatIOC> {
  return apiPost<ThreatIOC>('/api/threat-intel/iocs', input);
}

export function setThreatIOCWatchlist(id: string, watchlist: boolean): Promise<ThreatIOC> {
  return apiPatch<ThreatIOC>(`/api/threat-intel/iocs/${id}/watchlist`, { watchlist });
}

export function getThreatIntelSummary(): Promise<ThreatIntelSummary> {
  return apiGet<ThreatIntelSummary>('/api/threat-intel/summary');
}

File: backend/schemas/threat_intel.py
"""Request/response schemas for threat intelligence indicators."""
from typing import List, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from backend.schemas.common import Page, UtcDatetime

IOCType = Literal["ip", "domain", "url", "sha256"]
IOCSeverity = Literal["low", "medium", "high", "critical"]
IOCConfidence = Literal["low", "medium", "high"]


class RiskPoint(BaseModel):
    time: str = Field(min_length=1, max_length=40)
    score: int = Field(ge=0, le=100)


class ThreatIOCCreate(BaseModel):
    type: IOCType
    value: str = Field(min_length=1, max_length=512)
    severity: IOCSeverity
    confidence: IOCConfidence = "medium"
    description: str = Field(default="", max_length=2000)
    source: str = Field(min_length=1, max_length=200)
    first_seen: UtcDatetime
    last_seen: UtcDatetime
    watchlist: bool = False
    tags: List[str] = Field(default_factory=list, max_length=50)
    mitre_techniques: List[str] = Field(default_factory=list, max_length=50)
    risk_timeline: List[RiskPoint] = Field(default_factory=list, max_length=48)

    @field_validator("tags", "mitre_techniques")
    @classmethod
    def _bound_strings(cls, values: list[str]) -> list[str]:
        return [value.strip()[:120] for value in values if value.strip()]


class ThreatIOCWatchlistUpdate(BaseModel):
    watchlist: bool


class ThreatIOCItem(BaseModel):
    id: UUID
    type: IOCType
    value: str
    severity: IOCSeverity
    confidence: IOCConfidence
    description: str
    source: str
    first_seen: UtcDatetime
    last_seen: UtcDatetime
    watchlist: bool
    tags: List[str]
    mitre_techniques: List[str]
    risk_timeline: List[RiskPoint]
    created_at: UtcDatetime
    updated_at: UtcDatetime


class ThreatIOCPage(Page[ThreatIOCItem]):
    pass


class ThreatIntelSummary(BaseModel):
    total: int
    critical: int
    watchlist: int
    recent_48h: int
    items: List[ThreatIOCItem]
